### PR TITLE
added Tcp keepalive & removed Tcp heartbeat

### PR DESCRIPTION
**Pull request by phillyphil91**
## Current Behavior

Current implementations of TCP clients need to rely on ad-hoc heartbeats in order to keep the TCP connection open.

## Proposed Changes

Enable tcp_keepalive for the `TcpStream` through the `socket2` crate and remove usage of custom heartbeat.
This addresses: https://github.com/build-trust/ockam/issues/3450 and possibly https://github.com/build-trust/ockam/issues/3480

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

### CODE REVIEW

**Comment by phillyphil91** [comment]
The current changes "should" enable tcp keepalive for the `TcpStream`. Maybe it makes sense to add a test for this? On the other hand the only change is turning on tcp keepalive, so maybe that should already be tested by the `socket2` crate. 
Also I wasn't sure if the place where I instantiated the `socket2::TcpKeepalive` is the best one or if that should happen in the calling function `async fn process` in `listener.rs` or some other place.

I forgot to add my new key to git before the commit so the first commit is not verified. I'm not sure if I can fix this now. I can also add my name to the CONTRIBUTORS.csv file if this PR is merged.

File: implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs [hunk 1]
```diff
@@ -67,10 +65,12 @@ impl TcpSendWorker {
         stream: Option<TcpStream>,
         peer: SocketAddr,
         internal_addr: Address,
-        heartbeat: DelayedEvent<TcpSendWorkerMsg>,
     ) -> Self {
+        let keepalive = TcpKeepalive::new().with_time(Duration::from_secs(4));
         let (rx, tx) = match stream {
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: As far as I remember, there are 2 cases here: if we've been running a listener and just accepted new connection - in this case stream here is Some. In case we're initiating connection, stream is created in TcpSendWorker::initialize. According to this recently created issue #3480 we wanted to run keepalive only if we're initiating connection, but that was true for our custom keep-alive mechanism, let me clarify if that's also what we want with tcp-native keep-alive. I'll get back to you

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]
Looks great! Do you have any ideas on testing this feature? It's critical piece of functionality

File: implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs (hunk 1 above)
Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: Not entirely sure how to test this. Maybe it makes sense to set up a connection, don't send any data and then just use some crate that can monitor network traffic, if there are packets exchanged on the connection. Or maybe establish a connection and then sleep for some time before sending something to verify that the connection is still active.

I could take a look into this. Would you suggest adding it to the `ockam/implementations/rust/ockam/ockam_transport_tcp/tests` directory or maybe even to `send_receive.rs`?
There is an interesting function in there called `tcp_lifecycle__reconnect__should_not_error`. Maybe it make sense to test the tcp keepalive through that somehow.

Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: @phillyphil91 maybe you can connect to the cloud node, wait to X minutes (5-10-15?) and check that without keepalive the connection used to be dropped, while with keepalive you can still send a message to an Echoer and receive it back. This test should not be run by default though, we can't slow our CI this much

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: Yes I can do that. I might have some questions regarding the setup for the test along the way but I will see what I can do. 

I can add a flag that this is only tested when explicitly included.

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: @SanjoDeundiak I'm not really sure what you mean by cloud node?

Looking at the code right now, I have some questions regarding the test.
Would it make sense to use the setup of the `send_receive` function in  `ockam_transport_tcp/tests/send_receive.rs`?:

image
Although I am really not sure that I understand how the Sender is setup. 
Does this line basically mean that we are creating a tcp sender: 
`let r = route![(TCP, listener_address.to_string()), "echoer"];` ?

So fist we create a `TcpTransport` and then call listen() which returns a `SocketAddr` which is then passed in a tuple alongside `TCP` to `route!`, which does what exactly with the two values in the tuple? Is `TCP` not just a kind of `TransportType`?

The  result of all this is that we have a route of `1#127.0.0.1:8085 => 0#echoer` meaning we are sending a message to a `TcpSendWorker` which then sends the message to an echoer worker (which is not tcp)? 

Basically we want the `TcpSendWorker` to send keepalive packets. Does this mean i have to setup another `TcpSendWorker` because I think I have not fully understood how exactly the `send_receive` test works exactly and if this is really the setup where the keepalive packets should actually be exchanged.

Sorry for the many questions and if they are not particularly well formulated.

Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: Hey @phillyphil91 , your questions make perfect sense! Let me address them:
1. `route!` doesn't really do anything other than create a `Route` structure. Creating a route with `TCP` address in it doesn't really do anything. But when we send a message using that address, it first gets routed to our `TcpRouter` (because central Router knows `TCP` address type goes to `TcpRouter`), after that `TcpRouter` should figure out that we either have existing connecting to that tcp address and route the message to corresponding `TcpSendWorker`, or otherwise first create one and send that message to it.
2. This `send_receive` actually sends messages from a node to itself using loopback tcp connection, that is not very useful in our case (should we even test keep-alive for loopbacks? o_O)
3. We want to test keep-alive with our deployed cloud nodes (they're actually using elixir). In this case node with host `1.node.ockam.network` should be fine. It should have `Echoer` worker running on address `echo`. So the test could perform following:
    a. Send a message to the cloud node's echoer and check if we get it back, the route is `route![(TCP, "1.node.ockam.network:4000"), "echo"];`
    b. Wait for some long period of time (5-10-15 minutes)
    c. Do `a.` again, verify it's succeeding. 
I'm a little bit afraid that it won't catch situation when connection died, but is recreated while we send the message. But the plan above is a great start IMO!

Also, to my previous comment: we want to initiate KeepAlive only if we're the one who creates the connection (i.e. we're connecting, not listening)

Anything else I can help you with?

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: Hi @SanjoDeundiak thanks for your help. It took me a second but I finally understood, what needs to be tested. I still wasn't sure what you meant by the cloud node and if that was something that I needed to setup in the code. But after some investigating I found out that `1.node.ockam.network:4000` is actually a deployed ockam node on an ec2 instance (?) that has an echo worker running on it and that is available to connect to. Not sure why I didn't get that earlier.
Anyways I have added a test that connects to it, waits for 10 minutes (or even less)  and then tries to send another message. Unfortunately it is not working as expected because the connection is lost after some time and hence I assume the tcp_keepalive is not working properly:
Screenshot from 2022-10-03 00-39-31
While the "old" version with the heartbeats worked and the connection is not closed:
Screenshot from 2022-10-03 00-29-23

I tried adding further options to the the tcp_keepalive option to prevent the connection from being dropped, but nothing has worked thus far: 
image
(btw. as you can see in the test I had to set the timeout feature of the test to some big number because otherwise I always got the custom error message that the test timed out).

I will have to further investigate this and play around with the options a bit more. Maybe there is also something that I am not understanding correctly about TCP sockets and keepalive that prevents this from working.

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: Hi @SanjoDeundiak, sorry this is taking so long. So i realized that my test was using some function that ends up calling `TcpSendWorker::new_pair()` without passing it `Some(Stream)`. That is only done by `TcpListenProcessor.process()` (and then I can see that tcp keepalive is indeed enabled for the port).

But if I understood you correctly that is not what we want, i.e. we don't want the keepalive option to be enabled when we are listening, but only if we are initiating the connection. But the only place where this gets called is through `TcpRouter.handle_connect()` which calls `TcpSendWorker::start_pair` but passes `None` to the stream parameter.

That means there needs to be some stream created for the connection initiation side as well and not just for the listener side as it is in `TcpListenProcessor.process` in listener.rs:
```rust
async fn process(&mut self, ctx: &mut Self::Context) -> Result<bool> {
        debug!("Waiting for incoming TCP connection...");

        // Wait for an incoming connection
        let (stream, peer) = self.inner.accept().await.map_err(TransportError::from)?;
        debug!("TCP connection accepted");

        let handle_clone = self.router_handle.async_try_clone().await?;
        // And create a connection worker for it
        let (worker, pair) =
            TcpSendWorker::new_pair(handle_clone, Some(stream), peer, Vec::new()).await?;
```
We can see here that stream is created by calling accept on the `TcpListener`. Maybe it's possible to call connect  on the `TcpRouterHandle`  or something like that and have it return a stream as well which could be used when creating a new `TcpSendWorker` not through a listener. 

Does that make any sense to you? Or do you have any suggestions for this?

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: Hi @SanjoDeundiak or @mrinalwadhwa is there maybe any comment on this? I think I can resolve this issue, but I am currently stuck at the points mentioned above and I am not really sure how to proceed. 

Thanks a lot for any guidance or comment

Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: @phillyphil91 sorry, I have been unavailable for some time! For connecting side TCP stream is created inside `TcpSendWorker::initialize` function, which is called by our runtime when that worker's run loop is about to start

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: hey @SanjoDeundiak thanks for the update and help. I think that was the missing piece to make it work. Not sure why i didn't see earlier that the keepalive option on the socket needed to be created on `TcpSendWorker::initialize`. 

Everything works fine now, I believe. When the tcp keepalive option is enabled for the tcp socket, the connection to ockam cloud node does not die as it would without the option. I also verified that keepailve is enabled for the connection via netsat (it is not enabled without the additional code)

A few things for clarification around configuring the tcp keepalive option and the test. Feel free to confirm or suggest changes:

TCP:
- The tcp keepalive probes will be send out after 5 minutes
- There will be 60 probes send out, each after 5 seconds adding a total of 300 seconds, i.e. keeping the connection alive for another 5 minutes

Test:
- After having send a message to the ockam cloud node the thread sleeps for 350 seconds and the tries to send another message
- Tests timeout after 30 seconds so I added a timeout time of 400 seconds so the test will run completely
- I added a `#[ignore]` flag to the test to require explicitly running that test only if necessary to not hold up the CI pipeline

Let me know if that works for you or if there is anything else that needs to be changed. 

Maybe one more question: When sending the first message to the ockam node we create a tcp connection to it from our node and that is the same one that is being used after the threads sleeps, right? I mean otherwise the tcp keepalive code would not run the way it does and there wouldn't be the info message if tcp keepalive is not enabled:
image
But I just wanted to confirm that assumption.

Comment (phillyphil91) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: @SanjoDeundiak I just realized that since the tcp keepalive is enabled in the `TcpSendWorker::initialize` only, the Heartbeat implementation that was there before can remain at the other places in the code, right? Should i add it to the other code, e.g.`TcpSendWorker::new` etc. again where it was before? Or do you not want the Heartbeat anywhere anymore?

**Comment by mrinalwadhwa** [comment]
Since today is the last day of `hacktoberfest`, I've add the `hacktoberfest-accepted` label in case you're participating. We can merge the PR after the current on going discussion is resolved.

File: implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs [hunk 2]
```diff
@@ -93,3 +95,51 @@ async fn tcp_lifecycle__reconnect__should_not_error(ctx: &mut Context) -> Result
...
+    .await?;
+
+    // Wait to receive an echo and print it.
+    let reply = ctx.receive::<String>().await?;
+    info!("Sender has received the following echo: {}\n", reply);
+
+    // Sleep the thread to allow the tcp socket to send keepalive probes
+    let sleep_duration = Duration::from_secs(350);
+    info!("Slepping thread now for {:?}", sleep_duration);
+    thread::sleep(sleep_duration);
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs]: I think we don't need to hang the thread. How about replacing that with `ctx.sleep()` to use more async-friendly sleep?

File: implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs [hunk 3]
```diff
@@ -93,3 +95,51 @@ async fn tcp_lifecycle__reconnect__should_not_error(ctx: &mut Context) -> Result
...
+    )
+    .await?;
+
+    // Wait to receive an echo and print it.
+    let reply = ctx.receive::<String>().await?;
+    info!("Sender has received the following echo: {}\n", reply);
+
+    // Sleep the thread to allow the tcp socket to send keepalive probes
+    let sleep_duration = Duration::from_secs(350);
+    info!("Slepping thread now for {:?}", sleep_duration);
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs]: typo

File: implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs (hunk 1 above)
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:69]: @phillyphil91 I think we don't need Heartbeats in TCP anymore.

File: implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs [hunk 4]
```diff
@@ -177,6 +155,14 @@ impl Worker for TcpSendWorker {
                     return Err(TransportError::from(e).into());
                 }
             };
+
+            let keepalive = TcpKeepalive::new()
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:159]: Let's use following values for now:
tcp_keepalive_time = 300 (seconds)
tcp_keepalive_intvl = 75 (seconds)
tcp_keepalive_probes = 2 (number of probes)

**Comment by SanjoDeundiak** [comment]
Hey @phillyphil91 , very nice job! Could you please resolve my latest comments and do a rebase? Also, could you please update `Cargo.lock` and squash your changes into 1 commit that follows the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages)? That is required to pass CI checks

**Comment by phillyphil91** [comment]
hey @SanjoDeundiak. All done, please confirm .

**Review by SanjoDeundiak — DISMISSED** [review]

**Comment by SanjoDeundiak** [comment]
@phillyphil91 could you please sign the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

**Comment by phillyphil91** [comment]
@SanjoDeundiak  done

**Review by SanjoDeundiak — APPROVED** [review]

**Comment by mrinalwadhwa** [comment]
@phillyphil91 congratulations on an amazing first contribution 🥳🥳🥳🥳

**Comment by phillyphil91** [comment]
@mrinalwadhwa thanks. great experience.  hopefully more to come
